### DISCUSSION TITLE
Autocomplete for function definition using a TypeAlias is broken

### DISCUSSION THREAD

**Discussion by davidfokkema** [microsoft/pylance-release#6172] [Enhancement]
## Environment data

-   Language Server version: 2024.2.1 (pyright version 1.1.348, commit cfb1de0c)
-   OS and version: MacOS 14.2.1
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.1 (from conda environment)

## Code Snippet

Consider this definition:
```python
from typing import TypeAlias


Thing: TypeAlias = list[int]

class Base:

    def the_thing(self) -> Thing:
        return []
```

## Repro Steps

1. Start adding a subclass:
```python
class Child(Base):

    def the_thi
```
2. Then, type 'n'.
3. Note that the floating window shows that the return type is `Thing`. <img width="876" alt="Screenshot 2024-02-13 at 12 59 45" src="https://github.com/microsoft/pylance-release/assets/917137/81888e49-2bfe-40db-9c96-0e5e073448a6">
4. Hit Tab for autocomplete.
5. Autocomplete expands the TypeAlias to `list[int]` instead of `Thing`.

## Expected behavior

Expected output:
```python
class Child(Base):

    def the_thing(self) -> Thing:
        return super().the_thing()
```

## Actual behavior

```python
class Child(Base):

    def the_thing(self) -> list[int]:
        return super().the_thing()
```

## Logs

Could not create the issue with the logs; they were too long although I cleared the log output, enabled the trace setting, pressed n and tab and copied the resulting logs. Not sure what part I'll have to paste from the logs...

**Comment by davidfokkema** [comment]
Thanks to @davep for coming up with the MRE code.

**Comment by StellaHuang95** [comment]
Thank you for reporting this issue. Our current design is to print the original return type instead of the type alias for overridden methods. I don't consider it a bug, but tagging @heejaechang  to see if he has a different perspective.

**Comment by davidfokkema** [comment]
Thanks, @StellaHuang95! Is there a rationale for this? The tooltip that appears when you are in the process of overriding a method _does_ show the type alias. When you subclass something, you're probably not interested in returning the original type, but in the type alias. You may not even know or understand what the original type is.

**Comment by davidfokkema** [comment]
Ran into this problem a few more times while creating subclasses. Whenever I override `__init__()`, tab completion gives me an elaborate method definition which is unwanted. I get this:
```python
def __init__(
        self,
        *content: Option | Separator | None | ConsoleRenderable | RichCast | str,
        tooltip: ConsoleRenderable | RichCast | str | None = None,
    ):
```
while this is the original class definition:
```python
    def __init__(
        self,
        *content: NewOptionListContent,
        tooltip: RenderableType | None = None,
    ):
```
(I removed some other parameters which were simple types and thus unchanged.) Since I just want to subclass something, I don't want the type definition to change. That confuses my users...

**Comment by StellaHuang95** [comment]
Thanks for the clarification. We could potentially add a setting to control the printed return type to be either original type or type alias. Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by tallen12** [comment]
I have been running into this issue as well and came across this discussion while trying to fix it. Was this setting ever implemented or is there some other workaround implemented in the meantime?
